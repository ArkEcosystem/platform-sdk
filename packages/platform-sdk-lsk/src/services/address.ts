import { Exceptions, IoC, Services } from "@arkecosystem/platform-sdk";
import { getAddressFromPassphrase, getAddressFromPublicKey } from "@liskhq/lisk-cryptography";
import { utils } from "@liskhq/lisk-transactions";

@IoC.injectable()
export class AddressService extends Services.AbstractAddressService {
	public async fromMnemonic(
		mnemonic: string,
		options?: Services.IdentityOptions,
	): Promise<Services.AddressDataTransferObject> {
		try {
			return { type: "bip39", address: getAddressFromPassphrase(mnemonic) };
		} catch (error) {
			throw new Exceptions.CryptoException(error);
		}
	}

	public async fromPublicKey(
		publicKey: string,
		options?: Services.IdentityOptions,
	): Promise<Services.AddressDataTransferObject> {
		try {
			return { type: "bip39", address: getAddressFromPublicKey(publicKey) };
		} catch (error) {
			throw new Exceptions.CryptoException(error);
		}
	}

	public async validate(address: string): Promise<boolean> {
		try {
			utils.validateAddress(address);

			return true;
		} catch {
			return false;
		}
	}
}
