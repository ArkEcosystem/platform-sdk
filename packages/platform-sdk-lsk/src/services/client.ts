import { Coins, Contracts, Exceptions, Helpers } from "@arkecosystem/platform-sdk";
import { Arr } from "@arkecosystem/platform-sdk-support";

import { WalletData } from "../dto";
import * as DTO from "../dto";

export class ClientService implements Contracts.ClientService {
	readonly #http: Contracts.HttpClient;
	readonly #peer: string;

	readonly #broadcastErrors: Record<string, string> = {
		"Invalid sender publicKey": "ERR_INVALID_SENDER_PUBLICKEY",
		"Account does not have enough LSK": "ERR_INSUFFICIENT_FUNDS",
		"Sender does not have a secondPublicKey": "ERR_MISSING_SECOND_PUBLICKEY",
		"Missing signSignature": "ERR_MISSING_SIGNATURE",
		"Sender is not a multisignature account": "ERR_MISSING_MULTISIGNATURE",
	};

	private constructor({ http, peer }) {
		this.#http = http;
		this.#peer = peer;
	}

	public static async construct(config: Coins.Config): Promise<ClientService> {
		try {
			return new ClientService({
				http: config.get<Contracts.HttpClient>("httpClient"),
				peer: config.get<string>("peer"),
			});
		} catch {
			return new ClientService({
				http: config.get<Contracts.HttpClient>("httpClient"),
				peer: `${Arr.randomElement(config.get<Coins.CoinNetwork>("network").hosts)}/api`,
			});
		}
	}

	public async destruct(): Promise<void> {
		//
	}

	public async transaction(id: string): Promise<Contracts.TransactionData> {
		const result = await this.get("transactions", { id });

		return Helpers.createTransactionDataWithType(result.data[0], DTO);
	}

	public async transactions(
		query: Contracts.ClientTransactionsInput,
	): Promise<Contracts.CollectionResponse<Coins.TransactionDataCollection>> {
		const result = await this.get("transactions", query);

		return {
			meta: result.meta,
			data: Helpers.createTransactionDataCollectionWithType(result.data, DTO),
		};
	}

	public async wallet(id: string): Promise<Contracts.WalletData> {
		const result = await this.get("accounts", { address: id });

		return new WalletData(result.data[0]);
	}

	public async wallets(
		query: Contracts.ClientWalletsInput,
	): Promise<Contracts.CollectionResponse<Coins.WalletDataCollection>> {
		const result = await this.get("accounts", query);

		return {
			meta: result.meta,
			data: new Coins.WalletDataCollection(result.data.map((wallet) => new WalletData(wallet))),
		};
	}

	public async delegate(id: string): Promise<Contracts.WalletData> {
		const result = await this.get("delegates", { username: id });

		return new WalletData(result.data[0]);
	}

	public async delegates(
		query?: Contracts.KeyValuePair,
	): Promise<Contracts.CollectionResponse<Coins.WalletDataCollection>> {
		const result = await this.get("delegates");

		return {
			meta: result.meta,
			data: new Coins.WalletDataCollection(result.data.map((wallet) => new WalletData(wallet))),
		};
	}

	public async votes(
		id: string,
		query?: Contracts.KeyValuePair,
	): Promise<Contracts.CollectionResponse<Coins.TransactionDataCollection>> {
		throw new Exceptions.NotImplemented(this.constructor.name, "votes");
	}

	public async voters(
		id: string,
		query?: Contracts.KeyValuePair,
	): Promise<Contracts.CollectionResponse<Coins.WalletDataCollection>> {
		throw new Exceptions.NotImplemented(this.constructor.name, "voters");
	}

	public async syncing(): Promise<boolean> {
		throw new Exceptions.NotImplemented(this.constructor.name, "syncing");
	}

	public async broadcast(transactions: Contracts.SignedTransaction[]): Promise<Contracts.BroadcastResponse> {
		const result: Contracts.BroadcastResponse = {
			accepted: [],
			rejected: [],
			errors: {},
		};

		for (const transaction of transactions) {
			const { data, errors } = await this.post("transactions", transaction);

			if (data) {
				result.accepted.push(transaction.id);
			}

			if (errors) {
				result.rejected.push(transaction.id);

				if (!Array.isArray(result.errors[transaction.id])) {
					result.errors[transaction.id] = [];
				}

				for (const [key, value] of Object.entries(this.#broadcastErrors)) {
					if (errors[0].message.includes(key)) {
						result.errors[transaction.id].push(value);
					}
				}
			}
		}

		return result;
	}

	private async get(path: string, query?: Contracts.KeyValuePair): Promise<Contracts.KeyValuePair> {
		const response = await this.#http.get(`${this.#peer}/${path}`, query);

		return response.json();
	}

	private async post(path: string, body: Contracts.KeyValuePair): Promise<Contracts.KeyValuePair> {
		const response = await this.#http.post(`${this.#peer}/${path}`, body);

		return response.json();
	}
}
