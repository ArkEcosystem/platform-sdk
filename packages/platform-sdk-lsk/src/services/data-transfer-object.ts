import { Coins, Contracts, Helpers, Services } from "@arkecosystem/platform-sdk";

import * as DTO from "../dto";

export class DataTransferObjectService extends Services.AbstractDataTransferObjectService {
	private constructor(private decimals: string) {
		super();
	}

	public static async __construct(config: Coins.Config): Promise<DataTransferObjectService> {
		return new DataTransferObjectService(config.get(Coins.ConfigKey.CurrencyDecimals));
	}

	public signedTransaction(identifier: string, signedData: string): Contracts.SignedTransactionData {
		return new DTO.SignedTransactionData(identifier, signedData, signedData, this.decimals);
	}

	public transaction(transaction: unknown): Contracts.TransactionDataType {
		return Helpers.createTransactionDataWithType(transaction, DTO).withDecimals(this.decimals);
	}
}
