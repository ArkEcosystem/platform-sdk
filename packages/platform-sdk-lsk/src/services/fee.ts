import { Coins, Contracts } from "@arkecosystem/platform-sdk";
import * as transactions from "@liskhq/lisk-transactions";
import * as transactionsBeta from "@liskhq/lisk-transactions-new";

import { manifest } from "../manifest";

export class FeeService implements Contracts.FeeService {
	readonly #network;

	private constructor(network: Coins.CoinNetwork) {
		this.#network = network.crypto.networkId;
	}

	public static async construct(config: Coins.Config): Promise<FeeService> {
		return new FeeService(config.get<Coins.CoinNetwork>("network"));
	}

	public async destruct(): Promise<void> {
		//
	}

	public async all(days: number): Promise<Contracts.TransactionFees> {
		return {
			// Core
			transfer: this.transform("TRANSFER_FEE"),
			secondSignature: this.transform("SIGNATURE_FEE"),
			delegateRegistration: this.transform("DELEGATE_FEE"),
			vote: this.transform("VOTE_FEE"),
			multiSignature: this.transform("MULTISIGNATURE_FEE"),
			ipfs: this.transform(0),
			multiPayment: this.transform(0),
			delegateResignation: this.transform(0),
			htlcLock: this.transform(0),
			htlcClaim: this.transform(0),
			htlcRefund: this.transform(0),
			// Magistrate
			businessRegistration: this.transform(0),
			businessResignation: this.transform(0),
			businessUpdate: this.transform(0),
			bridgechainRegistration: this.transform(0),
			bridgechainResignation: this.transform(0),
			bridgechainUpdate: this.transform(0),
		};
	}

	private transform(type: string | number): Contracts.TransactionFee {
		let fee: number = type === 0 ? 0 : transactions.constants[type];

		if (this.#network === manifest.networks.betanet.crypto.networkId) {
			fee = transactionsBeta.constants[type];
		}

		return {
			static: `${fee}`,
			max: `${fee}`,
			min: `${fee}`,
			avg: `${fee}`,
		};
	}
}
