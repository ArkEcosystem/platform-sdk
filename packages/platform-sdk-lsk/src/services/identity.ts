import { Contracts, Exceptions } from "@arkecosystem/platform-sdk";
import * as cryptography from "@liskhq/lisk-cryptography";

export class IdentityService implements Contracts.IdentityService {
	public constructor(network: string) {
		//
	}

	public async getAddress(opts: Contracts.KeyValuePair): Promise<string> {
		if (opts.passphrase) {
			return cryptography.getAddressFromPassphrase(opts.passphrase);
		}

		if (opts.multiSignature) {
			throw new Exceptions.NotSupported(this.constructor.name, "getAddress#multiSignature");
		}

		if (opts.publicKey) {
			return cryptography.getAddressFromPublicKey(opts.publicKey);
		}

		if (opts.privateKey) {
			throw new Exceptions.NotSupported(this.constructor.name, "getAddress#privateKey");
		}

		if (opts.wif) {
			throw new Exceptions.NotSupported(this.constructor.name, "getAddress#wif");
		}

		throw new Error("No input provided.");
	}

	public async getPublicKey(opts: Contracts.KeyValuePair): Promise<string> {
		if (opts.passphrase) {
			return cryptography.getPrivateAndPublicKeyFromPassphrase(opts.passphrase).publicKey;
		}

		if (opts.multiSignature) {
			throw new Exceptions.NotSupported(this.constructor.name, "getPublicKey#multiSignature");
		}

		if (opts.wif) {
			throw new Exceptions.NotSupported(this.constructor.name, "getPublicKey#wif");
		}

		throw new Error("No input provided.");
	}

	public async getPrivateKey(opts: Contracts.KeyValuePair): Promise<string> {
		if (opts.passphrase) {
			return cryptography.getPrivateAndPublicKeyFromPassphrase(opts.passphrase).privateKey;
		}

		if (opts.wif) {
			throw new Exceptions.NotSupported(this.constructor.name, "getPrivateKey#wif");
		}

		throw new Error("No input provided.");
	}

	public async getWIF(opts: Contracts.KeyValuePair): Promise<string> {
		if (opts.passphrase) {
			throw new Exceptions.NotSupported(this.constructor.name, "getWIF#passphrase");
		}

		throw new Error("No input provided.");
	}

	public async getKeyPair(opts: Contracts.KeyValuePair): Promise<Contracts.KeyPair> {
		if (opts.passphrase) {
			const keyPair = cryptography.getPrivateAndPublicKeyFromPassphrase(opts.passphrase);

			return { publicKey: keyPair.publicKey, privateKey: keyPair.privateKey };
		}

		if (opts.publicKey) {
			throw new Exceptions.NotSupported(this.constructor.name, "getKeyPair#publicKey");
		}

		if (opts.privateKey) {
			throw new Exceptions.NotSupported(this.constructor.name, "getKeyPair#privateKey");
		}

		if (opts.wif) {
			throw new Exceptions.NotSupported(this.constructor.name, "getKeyPair#wif");
		}

		throw new Error("No input provided.");
	}
}
