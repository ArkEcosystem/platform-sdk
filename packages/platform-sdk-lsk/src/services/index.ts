export * from "./address";
export * from "./client";
export * from "./fee";
export * from "./key-pair";
export * from "./ledger";
export * from "./message";
export * from "./private-key";
export * from "./public-key";
export * from "./transaction";
