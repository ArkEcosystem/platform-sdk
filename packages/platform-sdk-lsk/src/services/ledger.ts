import { Coins, Contracts, Exceptions } from "@arkecosystem/platform-sdk";
import { BIP44 } from "@arkecosystem/platform-sdk-support";
import LedgerTransport from "@ledgerhq/hw-transport-node-hid-singleton";
import { CommHandler, DposLedger, LedgerAccount, SupportedCoin } from "dpos-ledger-api";

export class LedgerService implements Contracts.LedgerService {
	#ledger: LedgerTransport;
	#transport!: DposLedger;

	private constructor(transport: Contracts.LedgerTransport) {
		this.#ledger = transport;
	}

	public static async construct(config: Coins.Config): Promise<LedgerService> {
		try {
			return new LedgerService(config.get("services.ledger.transport"));
		} catch {
			return new LedgerService(LedgerTransport);
		}
	}

	public async destruct(): Promise<void> {
		await this.disconnect();
	}

	public async connect(): Promise<void> {
		this.#ledger = await this.#ledger.open();
		// @ts-ignore
		this.#transport = new DposLedger(new CommHandler(this.#ledger));
	}

	public async disconnect(): Promise<void> {
		await this.#ledger.close();
	}

	public async getVersion(): Promise<string> {
		const { version } = await this.#transport.version();

		return version;
	}

	public async getPublicKey(path: string): Promise<string> {
		const { publicKey } = await this.#transport.getPubKey(this.getLedgerAccount(path));

		return publicKey;
	}

	public async signTransaction(path: string, payload: Buffer): Promise<string> {
		const signature: Buffer = await this.#transport.signTX(this.getLedgerAccount(path), payload);

		return signature.toString("hex");
	}

	public async signTransactionWithSchnorr(path: string, payload: Buffer): Promise<string> {
		throw new Exceptions.NotImplemented(this.constructor.name, "signTransactionWithSchnorr");
	}

	public async signMessage(path: string, payload: Buffer): Promise<string> {
		const signature: Buffer = await this.#transport.signMSG(this.getLedgerAccount(path), payload);

		return signature.slice(0, 64).toString("hex");
	}

	public async signMessageWithSchnorr(path: string, payload: Buffer): Promise<string> {
		throw new Exceptions.NotImplemented(this.constructor.name, "signMessageWithSchnorr");
	}

	private getLedgerAccount(path: string): LedgerAccount {
		return new LedgerAccount().coinIndex(SupportedCoin.LISK).account(BIP44.parse(path).account);
	}
}
