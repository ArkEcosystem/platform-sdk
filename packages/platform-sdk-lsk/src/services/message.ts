import { Contracts } from "@arkecosystem/platform-sdk";
import * as cryptography from "@liskhq/lisk-cryptography";

export class MessageService implements Contracts.MessageService {
	public async sign(input): Promise<Contracts.SignedMessage> {
		return cryptography.signMessageWithPassphrase(input.message, input.passphrase);
	}

	public async verify(input): Promise<boolean> {
		return cryptography.verifyMessageWithPublicKey(input);
	}
}
