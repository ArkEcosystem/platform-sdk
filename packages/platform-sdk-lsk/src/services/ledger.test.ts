import "jest-extended";
import { createTransportReplayer, RecordStore } from "@ledgerhq/hw-transport-mocker";
import { LedgerService } from "../../src/services/ledger";
import { ledger } from "../__fixtures__/ledger";
import { createConfig } from "../helpers";

const createMockService = async (record: string) => {
	const transport = await LedgerService.construct(createConfig());

	await transport.connect(createTransportReplayer(RecordStore.fromString(record)));

	return transport;
};

describe("destruct", () => {
	it("should pass with a resolved transport closure", async () => {
		const lsk = await createMockService("");

		await expect(lsk.destruct()).resolves.toBeUndefined();
	});
});

describe("getVersion", () => {
	it("should pass with an app version", async () => {
		const lsk = await createMockService(ledger.appVersion.record);

		await expect(lsk.getVersion()).resolves.toEqual(ledger.appVersion.result);
	});
});

describe("getPublicKey", () => {
	it("should pass with a compressed publicKey", async () => {
		const lsk = await createMockService(ledger.publicKey.record);

		await expect(lsk.getPublicKey(ledger.bip44.path)).resolves.toEqual(ledger.publicKey.result);
	});
});

describe("signTransaction", () => {
	it("should pass with a signature", async () => {
		const lsk = await createMockService(ledger.transaction.record);

		await expect(
			lsk.signTransaction(ledger.bip44.path, Buffer.from(ledger.transaction.payload, "hex")),
		).resolves.toEqual(ledger.transaction.result);
	});
});

describe("signTransactionWithSchnorr", () => {
	it("should fail with a 'NotImplemented' error", async () => {
		const lsk = await createMockService("");

		await expect(lsk.signTransactionWithSchnorr("", Buffer.alloc(0))).rejects.toThrow();
	});
});

describe("signMessage", () => {
	it("should pass with a signature", async () => {
		const lsk = await createMockService(ledger.message.record);

		await expect(lsk.signMessage(ledger.bip44.path, Buffer.from(ledger.message.payload, "hex"))).resolves.toEqual(
			ledger.message.result,
		);
	});
});

describe("signMessageWithSchnorr", () => {
	it("should fail with a 'NotImplemented' error", async () => {
		const lsk = await createMockService("");

		await expect(lsk.signMessageWithSchnorr("", Buffer.alloc(0))).rejects.toThrow();
	});
});
