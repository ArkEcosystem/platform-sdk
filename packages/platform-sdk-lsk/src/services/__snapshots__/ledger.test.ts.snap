// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scan should return scanned wallet 1`] = `
Object {
  "m/44'/134'/0'/0/0": WalletData {
    "data": Object {
      "address": "7399986239080551550L",
      "asset": Object {},
      "balance": "7370000000",
      "publicKey": "d48522677df50defd175c85072309c7643dbc6bdc63c7665a302579ed2ccaedb",
      "secondPublicKey": "",
    },
  },
  "m/44'/134'/1'/0/0": WalletData {
    "data": Object {
      "address": "11603034586667438647L",
      "asset": Object {},
      "balance": "1180000000",
      "publicKey": "635ec78f25eefe8dcb7b4e7f5685c5aad1fef54048c596ee4e74cdf5e7b0f37c",
      "secondPublicKey": "",
    },
  },
  "m/44'/134'/2'/0/0": WalletData {
    "data": Object {
      "address": "8261766349562104742L",
      "asset": Object {},
      "balance": "700000000",
      "publicKey": "",
      "secondPublicKey": "",
    },
  },
  "m/44'/134'/3'/0/0": WalletData {
    "data": Object {
      "address": "10806468874187168404L",
      "balance": 0,
      "publicKey": "5c12c704e97b79a7abd0cdb36336e4bdc675fcec5b8844d1fc95380c1d5db052",
    },
  },
}
`;
